drivers: lpadc: enable nxp lpadc acquisition time feature

ZhaoxiangJin opened the pull request:
add nxp lpadc acquisition time feature.
Tested locally on evkmimxrt500, mimxrt1160, lpcxpress5536, whether the acquisition time type passed in is ticket, microsecond, or nanosecond, the register settings can be successfully performed.

ZhaoxiangJin committed:
drivers: clock_control: add support for LPADC clock obtain

The lpadc driver needs to obtain its functional clock to configure
the acquisition time. This patch add support for I.MX RT three digit
parts, I.MX RT11xx parts, and LPC parts.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
dts: nxp: Add LPADC clocks properities to SoC dtsi

Add LPADC clock properties to SoC dtsi on adc/lpadc node

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
drivers: lpadc/adc_mcux_lpadc/: enable nxp lpadc acquisition time feature

Enable nxp lpadc driver acquisition time feature

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

danieldegrasse commented on drivers/clock_control/clock_control_mcux_ccm.c:
nit here- commit title is misspelled:
`drivers: clock_control: add supoort for LPADC functional clock frequency `

Also, it looks like this commit will add support for iMX RT 3 digit parts, iMX RT11xx parts, and LPC parts. Is that right? If so, the body of this commit needs to be updated as well.

danieldegrasse commented on drivers/clock_control/clock_control_mcux_ccm.c:
```diff
@@ -377,6 +377,15 @@ static int mcux_ccm_get_subsys_rate(const struct device *dev,
 #endif
+
+#ifdef CONFIG_ADC_MCUX_LPADC
+	case IMX_CCM_LPADC1_CLK:
```
Is there a part using the CCM driver that has an LPADC?

ZhaoxiangJin replied:
Hi Daniel, the I.MX RT 11xx series use the CCM

danieldegrasse commented on include/zephyr/dt-bindings/clock/imx_ccm.h:
```diff
@@ -66,4 +66,7 @@
 #define IMX_CCM_FLEXIO1_CLK		0x1100UL
 #define IMX_CCM_FLEXIO2_3_CLK		0x1101UL
 
+#define IMX_CCM_LPADC1_CLK 0x1200UL
```
Please align indentation here (although if there is no SOC using the CCM driver that has an LPADC, this change can just be dropped)

danieldegrasse commented on include/zephyr/dt-bindings/clock/imx_ccm_rev2.h:
```diff
@@ -99,4 +99,8 @@
 #define IMX_CCM_PIT1_CLK                0x5001UL
 
+/* ADC */
+#define IMX_CCM_LPADC1_CLK 0x6000UL
```
As above- please align indentation

danieldegrasse commented on include/zephyr/dt-bindings/clock/mcux_lpc_syscon_clock.h:
```diff
@@ -74,4 +74,7 @@
 #define MCUX_PORT4_CLK			MCUX_LPC_CLK_ID(0x0C, 0x04)
 #define MCUX_PORT5_CLK			MCUX_LPC_CLK_ID(0x0C, 0x05)
 
+#define MCUX_LPADC1_CLK MCUX_LPC_CLK_ID(0x0D, 0x00)
```
As above- please align indentation

danieldegrasse commented on drivers/adc/adc_mcux_lpadc.c:
```diff
@@ -57,7 +59,108 @@ struct mcux_lpadc_data {
+	uint32_t acquisition_time_value = ADC_ACQ_TIME_VALUE(acq_time);
+	uint8_t acquisition_time_unit = ADC_ACQ_TIME_UNIT(acq_time);
+	uint32_t sample_time_array[] = {3, 5, 7, 11, 19, 35, 67, 131};
+	uint8_t sample_time_array_size = sizeof(sample_time_array) / sizeof(*sample_time_array);
```
You can probably remove this variable, and just use `ARRAY_SIZE(sample_time_array)` instead

danieldegrasse commented on drivers/adc/adc_mcux_lpadc.c:
```diff
@@ -57,7 +59,108 @@ struct mcux_lpadc_data {
+		} else if (sample_time_array[sample_time_array_size - 1] < acquisition_time_value) {
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK131;
+		} else {
+			while ((array_right_index - array_left_index) > 1) {
```
I think a comment describing this search algorithm might be useful. From what I can tell we are comparing the midpoint of all possible ADC acquisition times, and based on that we search the lower or upper half of the array, repeating this process until we find the best acquisition time.

It might also be worth mentioning in a comment that the indices in the array correspond to the register values that need to be written to select specific acquisition times

ZhaoxiangJin commented on include/zephyr/dt-bindings/clock/imx_ccm_rev2.h:
```diff
@@ -99,4 +99,8 @@
 
+/* ADC */
+#define IMX_CCM_LPADC1_CLK	0x6000UL
+#define IMX_CCM_LPADC2_CLK	0x6001UL
```
Hi @danieldegrasse, I know you enabled the FLEXSPI clock, I kept the indentation consistent with yours (FLEXSPI), but it seems inconsistent with the PIT.

danieldegrasse replied:
Sure- if you'd like to align the indentation that would be great. If possible, put the changes to fix indentation in a separate commit just to keep the history clean

ZhaoxiangJin committed:
drivers: clock_control: imx_ccm_rev2.h: Unified indentation format

Unified indentation format of imx_ccm_rev2.h file.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin commented:
Hi @anangl, could you please help review this PR, thanks.

anangl commented on drivers/adc/adc_mcux_lpadc.c:
```diff
@@ -57,7 +59,119 @@ struct mcux_lpadc_data {
+		if (ADC_ACQ_TIME_MICROSECONDS == acquisition_time_unit) {
+			acquisition_time_value *= 1000;
+		}
+		/* If the acquisition time that needs to be configured is less than the minimum
+		 * value that can be configured, configure the minimum value into the register. If
+		 * the acquisition time that needs to be configured is greater than the maximum
+		 * value that can be configured, configure the maximum value into the register.
+		 * Otherwise, compare the acquisition time that needs to be configured with the
+		 * midpoint of the array, and iterate the left or right boundary of the array based
+		 * on the comparison result until the most appropriate configuration option is
+		 * found.
```
I don't think this is the correct approach. You should rather take the closest configurable value that is not shorter than the requested acquisition time. And return -EINVAL if that requested time exceeds what can be configured.

ZhaoxiangJin replied:
> I don't think this is the correct approach. You should rather take the closest configurable value that is not shorter than the requested acquisition time. And return -EINVAL if that requested time exceeds what can be configured.

Thanks for your comments, I'd like to know why you think the configure value should be greater than the input acquisition time.

ZhaoxiangJin replied:
The following statement explains why I use the floating point value. I still do not know why **configure value should be greater than the input acquisition time** just like @anangl said. Seems this code caused trouble to many people, maybe it is enough to only support the acquisition time being passed in ADCK units? 

Why I think the variable adc_cycle_ns needs to be set to a floating point type, I would like to give an example to illustrate. Under NXP LPC55S36 platform, the default ADC clock frequency is 48MHz, so the ADC cycle is approximately 20.833 ns (if variable adc_cycle_ns is set to an integer type, then this number is 20).

case1(float point type, 20.833):
sample_time_array[0] = 20.833 * 3 = 62.499 ns
sample_time_array[1] = 20.833 * 5 = 104.165 ns
if the parameter ADC acquisition time passed by the application layer is 82ns, then the option kLPADC_SampleTimeADCK3 will be set to the corresponding bitfield.

case1(integer type, 20):
sample_time_array[0] = 20 * 3 = 60 ns
sample_time_array[1] = 20 * 5 = 100 ns
if the parameter ADC acquisition time passed by the application layer is 82ns, then the option kLPADC_SampleTimeADCK5 will be set to the corresponding bitfield.

If we use integer type in the code, the accuracy of the calculation results decreases. At this clock frequency (48MHz), and application configuration options (82ns), the optimal configuration item should be kLPADC_SampleTimeADCK3.

Currently, my local code has changed the ADC sample time array (sample_time_array) to the integer type. The reason is that converting the final calculation result into an integer will not cause too much loss of accuracy, and at the same time reduces the calculation pressure of chips without FPU platforms (the subsequent calculation process uses array elements to compare with the acquisition_time_value).

for example
sample_time_array[0] = 20.833 * 3 = 62.499 ns = 62
sample_time_array[1] = 20.833 * 5 = 104.165 ns = 104
if the parameter ADC acquisition time passed by the application layer is 82ns, then the option kLPADC_SampleTimeADCK3 will be set to the corresponding bitfield.

ZhaoxiangJin commented:
I changed the approach.
1. Do not use the floating point, @decsny please review.
2. Find the nearest larger value to configure, @anangl  please review.

decsny commented on drivers/adc/adc_mcux_lpadc.c:
```diff
@@ -57,7 +58,65 @@ struct mcux_lpadc_data {
+			acquisition_time_value *= 1000;
+		}
+
+		for (uint8_t index = 0; index < sample_time_array_size; ++index) {
```
the two for loops appear they can be combined into one

decsny commented on drivers/adc/adc_mcux_lpadc.c:
```diff
@@ -57,7 +58,65 @@ struct mcux_lpadc_data {
+				find_suitable_option = true;
+				break;
+			}
+		}
```
rather than using the stack array and the for loops, can't you just have a if/else if code and a multiplier variable?

decsny replied:
a loop brings extra code size too though, and these stack variables are also in the flash since they are literals, so that's not much different as far as code size, the issue is the stack size when having an array, the stack sizes on the embedded devices using zephyr will be limited and we should reduce the use of it

ZhaoxiangJin replied:
I feel confused about the **these stack variables are also in the flash since they are literals**, I checked the .map file, the stack is located in the RAM area. 

You are right, we should really pay attention to the use of stacks in embedded devices. But I think there is no big problem with the stack used in the code added by this PR, because this is a NOT big array used during the initialization process, this stack will also be released at the end of the function call, and the function call depth is very low.

My main consideration is that I don’t want to write code that is too hardcoded (directly embedding values ​​such as 3, 5, ... into the code), which will make it very troublesome to maintain the code later.

ZhaoxiangJin replied:
I will update the code with removing the array and using the if/else code instead of using for loop.

decsny commented on drivers/adc/adc_mcux_lpadc.c:
```diff
@@ -57,13 +58,70 @@ struct mcux_lpadc_data {
+			return -EINVAL;
+		}
+
+		conversion_factor = (uint32_t)((float)1000000000.0 / adc_freq_hz);
```
I know I asked this before, but now I still really don't get value of the intermediate conversion to float, since the max value here (1 billion, if I am reading the number correctly) is within the range of ints, and the fractional part will be truncated anyways by converting back to int, so there seems to be no increase in the range or precision by converting to float.... am I misunderstanding something

decsny commented on drivers/adc/adc_mcux_lpadc.c:
```diff
@@ -57,13 +58,70 @@ struct mcux_lpadc_data {
+		cmd->sampleTimeMode = kLPADC_SampleTimeADCK11;
+	} else if ((19 * conversion_factor) >= acquisition_time_value) {
+		cmd->sampleTimeMode = kLPADC_SampleTimeADCK19;
+	} else if ((35 * conversion_factor) >= acquisition_time_value) {
```
this code is fine now, I much prefer this approach, but one thing I remember you said is you didn't know why to just pick the greatest option rather than the closest.... well, I don't know the reason either, maybe @anangl can clarify, but I think if you want to pick the closest value, that seems fine to me.... the way you can do it, is to check if the value is greater than the midpoint between the points, rather than the points itself, like this:

```
if ((4* conversion_factor) >= time) {
mode = 3
} else if (6*fact >= time) {
mode = 5
} else if(9 *factor >= time {
mode = 7
} ... etc
```
see what I mean about how it will pick then the closest value ?

ZhaoxiangJin replied:
the way you provided to find the most closest options is also ok, but the time complexity of binary search is better than the ordered search. Of course, the amount of options is not large, so the performance of the two algorithms is similar.

And I guess @anangl may think we should ensure sufficient sampling time of the ADC, because the sampling time will affect the accuracy of the ADC, although this will result in a drop in sampling rate.

decsny reviewed (commented):
the PR is looking fine to me now just wanted your opinion about 2 comments I have left still

dleach02 commented:
@ZhaoxiangJin, please address the merge conflicts

ZhaoxiangJin committed:
samples: drivers: adc: Update acquisition time property configuration

Enabled nxp lpadc driver acquisition time feature, the overlay file
can express acquisition time in microseconds, nanosecond and ADC
ticks.

The purpose of updating these overlay files is not only for local
testing, but also to remind users that they can use this feature.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin commented:
Hi @anangl, please help to review, thanks.

ZhaoxiangJin commented:
Could anyone else help review this PR? Thanks very much.
